src/controllers/trackController.ts

```diff
@@ -181,15 +188,71 @@ const uploadPatchToS3 = asyncWrapper(
   }
 );
 
+const checkTrackOwner = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    const track = await Track.findById(req.params.id);
+
+    if (!track) {
+      return next(new AppError('No Track found with that ID ', 404));
+    }
+
+    const isOwner = checkDocumentOwner(track, req.user);
+
+    if (!isOwner)
+      return next(
+        new AppError('You are not authorized to update this document', 401)
+      );
+
+    req.track = track;
+
+    next();
+  }
+);
+
 const createTrack = createOne<ITrack>(Track);
 
 const getAllTracks = getAllEntities<ITrack>(Track);
 
 const getTrack = getOne<ITrack>(Track);
 
-const updateTrack = updateOne<ITrack>(Track, true);
+const updateTrack = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    const updatedTrack = await Track.findByIdAndUpdate(
+      req.params.id,
+      req.body,
+      { runValidators: true, new: true }
+    );
+
+    res.status(201).json({
+      status: 'success',
+      data: updatedTrack,
+    });
+
+    // Deleting old files from AWS S3
+    let deleteUrls: string[] = [];
+    const track = req.track;
 
-const deleteTrack = deleteOne<ITrack>(Track, true);
+    if (req.body.coverImage && track) deleteUrls.push(track.coverImage);
+    if (req.body.url && track) deleteUrls.push(track.url);
+
+    if (deleteUrls.length > 0) await removesFilesFromS3(deleteUrls);
+  }
+);
+
+const deleteTrack = asyncWrapper(
+  async (req: IRequestWithUser, res: Response, next: NextFunction) => {
+    await Track.findByIdAndRemove(req.params.id);
+
+    res.status(204).json({
+      status: 'success',
+      data: null,
+    });
+
+    // Deleting old files from AWS S3
+    const track = req.track;
+    if (track) await removesFilesFromS3([track.coverImage, track.url]);
```

Comment:
Why are you deleting the files after sending a response to the user? What happens if the deletion process fails?

Reply:
Oh that's true, you need to consider add validations and move the response to the final.

Reply:
That's the thing, even if the delete process fails, the entity will be deleted/updated from the DB, so, sending the response before would make sense because the delete process is totally outside of the user concern.

What I was thinking is that maybe add a collection that keeps track of failed deletion process and then afterwards an admin could check these documents and delete them manually

Reply:
Honestly, I don't like it, it is true that is not of user concern if the picture deletion is successful or not but it is still an operation performed when triggered by a user request so it should be performed before sending a response to the user or asynchronously in another process. This being a simple project I don't think you'd have problems performing the deletion before sending a user response but in a more "production ready" system you'd either have a different process in charge of deleting pictures or a different service doing the same.

Reply:
> That's the thing, even if the delete process fails, the entity will be deleted/updated from the DB, so, sending the response before would make sense because the delete process is totally outside of the user concern.
> 
> What I was thinking is that maybe add a collection that keeps track of failed deletion process and then afterwards an admin could check these documents and delete them manually

I was going to mention something similar to you, but since this is a practice project, I did not want to mention that it would be correct to have some records where it is stored if it fails, so that the case that I mentioned occurs, so that it does not run out of photos in that case. that the update fails and the photos are also deleted, or the deletion of the photos fails